components/feature/downloads/src/main/java/mozilla/components/feature/downloads/share/ShareDownloadFeature.kt

```diff
@@ -0,0 +1,209 @@
+import mozilla.components.support.ktx.android.content.shareMedia
+import mozilla.components.support.ktx.kotlin.sanitizeURL
+import mozilla.components.support.ktx.kotlinx.coroutines.flow.ifChanged
+import java.io.File
+import java.io.FileOutputStream
+import java.io.IOException
+import kotlin.math.absoluteValue
+import kotlin.random.Random
+
+/**
+ * Default mime type for base64 images data URLs not containing the media type.
+ */
+@VisibleForTesting
+internal const val DEFAULT_IMAGE_EXTENSION = "jpg"
+
+/**
+ * At most time to allow fo the file to be downloaded and the share intent chooser to appear.
+ */
+private const val OPERATION_TIMEOUT_MILLIS = 1000L
+
+/**
+ * Subdirectory of Context.getCacheDir() where the resources to be shared are stored.
+ *
+ * Location must be kept in sync with the paths our FileProvider can share from.
+ */
+@VisibleForTesting
+internal var cacherDirName = "mozac_share_cache"
+
+/**
+ * [Middleware] implementation for sharing online resources.
+ *
+ * This will intercept only [ShareInternetResourceAction] [BrowserAction]s.
+ *
+ * Following which it will transparently
+ *  - download internet resources while respecting the private mode related to cookies handling
+ *  - temporarily cache the downloaded resources
+ *  - automatically open the platform app chooser to share the cached files with other installed Android apps.
+ *
+ * with a 1 second timeout to ensure a smooth UX.
+ *
+ * To finish the process in this small timeframe the feature is recommended to be used only for images.
+ *
+ *  @param context Android context used for various platform interactions
+ *  @param httpClient Client used for downloading internet resources
+ *  @param store a reference to the application's [BrowserStore]
+ */
+@Suppress("TooManyFunctions")
+class ShareDownloadFeature(
+    private val context: Context,
+    private val httpClient: Client,
+    private val store: BrowserStore,
+    private val tabId: String?
+) : LifecycleAwareFeature {
+    private val logger = Logger("ShareDownloadMiddleware")
+
+    @VisibleForTesting
+    internal var scope: CoroutineScope? = null
+
+    override fun start() {
+        scope = store.flowScoped { flow ->
+            flow.mapNotNull { state -> state.findTabOrCustomTabOrSelectedTab(tabId) }
+                .ifChanged { it.content.share }
+                .collect { state ->
+                    state.content.share?.let { shareState ->
+                        logger.debug("Starting the sharing process")
+                        startSharing(shareState)
+
+                        // This is a fire and forget action, not something that we want lingering the tab state.
+                        store.dispatch(ShareInternetResourceAction.ConsumeShareAction(state.id))
+                    }
+                }
+        }
+    }
+
+    override fun stop() {
+        scope?.cancel()
+    }
+
+    init {
+        CoroutineScope(IO).launch {
+            cleanupCache()
+        }
```

Comment:
Could we add test to cover this use case?

Reply:
I really tried the first time but saw no easy way to especially since we are using `launch` here and I feel like it makes sense to use it so.
Did another try in which I observe the effects, after a short delay. Hope it's ok.

Reply:
I think one solution could be to hide the launch call inside `cleanupCache` and then stub the call to do nothing and verify that every time `init` is called we call `cleanupCache`.

Reply:
This would be nice but I don't know hot to stub the behavior before the object.
The way I see it to check for the right behavior I need a spy which will receive a constructed instance in which the init has already ran.
I'd appreciate any pointers on how to achieve this.